README.md
    ____   ____ ______ ____    ___
    || \\ ||    | || | || \\  // \\
    ||_// ||==    ||   ||_// ((   ))
    || \\ ||___   ||   || \\  \\_//
    a personal, minimalistic forth

[![Build Status](https://travis-ci.org/crcx/retro12.svg?branch=master)](https://travis-ci.org/crcx/retro12)

## Background

RETRO is a Forth dialect drawing influences from many sources. It uses
some, but not all of Chuck Moore's newer ideas. It's clean, elegant,
and tiny. It may not have many features, and it may not be particularly
useful by itself, but it's easy to grasp and easily adapted to various
uses.

This is the twelfth generation of the language and related programming
environment. It is a fresh start and is not compatible with the
previous generation.

## Quick Start

For Linux and macOS:

    make
    ./bin/kanga

## Legalities

Permission to use, copy, modify, and/or distribute this software for
any purpose with or without fee is hereby granted, provided that the
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.

Retro 12 is:

    Copyright (c) 2008 - 2017, Charles Childers

Portions of the code derive from Retro 11, which is:

    Copyright (c) 2008 - 2016, Charles Childers
    Copyright (c) 2012 - 2013, Michal J Wallace
    Copyright (c) 2009 - 2011, Luke Parrish
    Copyright (c) 2009 - 2010, JGL
    Copyright (c) 2010 - 2011, Marc Simpson
    Copyright (c) 2011 - 2012, Oleksandr Kozachuk
    Copyright (c) 2010,        Jay Skeer
    Copyright (c) 2010,        Greg Copeland
    Copyright (c) 2011,        Aleksej Saushev
    Copyright (c) 2011,        Foucist
    Copyright (c) 2011,        Erturk Kocalar
    Copyright (c) 2011,        Kenneth Keating
    Copyright (c) 2011,        Ashley Feniello
    Copyright (c) 2011,        Peter Salvi
    Copyright (c) 2011,        Christian Kellermann
    Copyright (c) 2011,        Jorge Acereda
    Copyright (c) 2011,        Remy Moueza
    Copyright (c) 2012,        John M Harrison
    Copyright (c) 2012,        Todd Thomas


CHANGELOG.txt

2017-08-24  Charles Childers  <crc@forthworks.com>

	* add set:reduce
	* next release will be 2017.9
	* add s:copy, document additions to s: namespace

2017-08-21  Charles Childers  <crc@forthworks.com>

	* some new constants in s: namespace
	* bump image to 2017.8
	* some tests for choose, n:negative?

2017-08-02  Charles Childers  <crc@forthworks.com>

	* tests for n:even?, n:odd?, times, n:between?
	* tests for n:min, n:max, n:negate
	* tests for n:limit, v:limit, tuck, fetch-next

2017-07-31  Charles Childers  <crc@forthworks.com>

	* add tests for gt?, lt?, gteq?, lteq?
	* add tests for c:vowel?, c:-vowel?
	* add tests for store, fetch
	* add tests for n:inc, n:dec, const
	* add tests for var, var<n>
	* add tests for v:dec, v:dec-by

2017-07-30  Charles Childers  <crc@forthworks.com>

	* add test for case, s:case
	* initial test for v:inc, v:preserve
	* add initial tests for tri, tri@, tri*
	* add tests for bi, bi@, bi*

2017-07-24  Charles Childers  <crc@forthworks.com>

	* add more tests
	* posix files: more commentary, add file:read-line, file:for-each-line
	* update CHANGELOG
	* tanu is now a filter; closes #12
	* more readability improvements to the Makefile generator
	* use the new build tool to generate a Makefile for the interfaces, extend, and embedimage

2017-07-22  Charles Childers  <crc@forthworks.com>

	* remove unused file
	* remove unused file
	* build.forth: cleaner Makefile output
	* fix listener script
	* new build tool now capable of outputting a valid Makefile
	* more work on new build tool: scan dependencies
	* build.forth: take argument from command line
	* build tool: inching closer
	* more work on build.forth
	* new example

2017-07-21  Charles Childers  <crc@forthworks.com>

	* start work on something like build.c, but in forth
	* begin early work to use new build tool for compiling sources
	* bring in new build tool (intended to help work on issue #13)
	* move documentation into doc dir
	* floating point now an io module
	* close block for floating point code
	* add a .gitattributes
	* seriously raise memory limits: 1M return stack; 4K stack depth; 48M heap
	* split rre.forth better
	* begin work to support embedding source files into rre, other interfaces

2017-07-20  Charles Childers  <crc@forthworks.com>

	* minor update to Primes example; now supports much bigger sets
	* add some more tests
	* Fix example. Closes issue #11
	* begin work on fixing argument handler. see issue #11
	* add a couple more tests

2017-07-19  Charles Childers  <crc@forthworks.com>

	* a few more tests
	* example: using VT100/ANSI escape sequences
	* fix issue w/documentation for s:with-format
	* use .forth suffix for examples, not .md
	* add note on the fenced syntax to Interfaces.md
	* new example: simple implementation of "echo"
	* add support for ~~~ code blocks as well. closes #10
	* begin work on fixing issue #10 (better support for fenced code blocks)
	* Merge pull request #9 from ratanasov/master

2017-07-19  Radoslav Atanasov  <rsrado@gmail.com>

	* Included <stdint.h> to make int32_t visible

2017-07-18  Charles Childers  <crc@forthworks.com>

	* a few more tests
	* new example
	* split legalese into separate LICENSE file
	* add 48 new tests
	* increase test coverage (all of ASCII:, some others)
	* remove some unused code
	* fix initial instance of unspec behaviour in bridge.c (see issue #8)

2017-07-17  Charles Childers  <crc@forthworks.com>

	* start bringing in tests from retro11
	* begin work on an automated testing set
	* document the file: namespace, fix stack comments in rre.forth.md
	* some additional comments on shifts in Nga documentation
	* more work on travis support
	* begin work on travis ci support
	* add sys: namespace, with argc/argv words (addresses issue #6)

2017-07-16  Charles Childers  <crc@forthworks.com>

	* correct f:to-number max value issue; fixes bug #5
	* Merge pull request #4 from ratanasov/master

2017-07-16  Radoslav Atanasov  <rsrado@gmail.com>

	* `@` and `!` added to the prefix table in `RetroForth.md`
	* `make clean` now deletes `bin/repl` too.

2017-07-12  Charles Childers  <crc@forthworks.com>

	* add n:MIN and n:MAX constant
	* fix for issue #3 (bug in n:to-string)

2017-07-11  Charles Childers  <crc@forthworks.com>

	* fix documentation issue
	* update changelog
	* documentation for FPU
	* include f: namespace by default
	* initial checkin supporting FPU (port from WIP on iOS implementation)
	* fix dump-stack

2017-07-08  Charles Childers  <crc@forthworks.com>

	* rename .s (in prep for using . as prefix for floating point)

2017-06-15  Charles Childers  <crc@forthworks.com>

	* add set:length
	* example: finding prime numbers

2017-06-14  Charles Childers  <crc@forthworks.com>

	* stdlib now includes a Retro implementation of muri assembler
	* bump image version
	* update changelog
	* merge in samples, documentation

2017-05-26  Charles Childers  <crc@forthworks.com>

	* documentation updates
	* add note on RRE to Interfaces
	* remove old files
	* rre now uses literate sources
	* muri now embeds unu functionality
	* start work on embedding unu functionality into tools directly

2017-05-18  Charles Childers  <crc@forthworks.com>

	* update changelog
	* move depth earlier to fix a bug
	* add set: to words.txt
	* add set: vocabulary for simple arrays

2017-05-15  Charles Childers  <crc@forthworks.com>

	* add .s

2017-05-12  Charles Childers  <crc@forthworks.com>

	* add v:preserve

2017-05-11  Charles Childers  <crc@forthworks.com>

	* revert the multiple entry points - proved buggy
	* start using fall-through; some cleanups
	* new image w/updated prefix::
	* prefix:: now allows for multiple entry points

2017-05-01  Charles Childers  <crc@forthworks.com>

	* fix some stuff in the Makefile
	* add description of the interfaces
	* add kanga, a full screen curses-based interactive environment
	* add bridge.h, separate compile of bridge.c

2017-04-30  Charles Childers  <crc@forthworks.com>

	* add a changelog


source/naje.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include "nga.c"
#ifndef strtok_r
char* strtok_r(char *str, const char *delim, char **nextp) {
  char *ret;
  if (str == NULL) {
    str = *nextp;
  }
  str += strspn(str, delim);
  if (*str == '\0') {
    return NULL;
  }
  ret = str;
  str += strcspn(str, delim);
  if (*str) {
    *str++ = '\0';
  }
  *nextp = str;
  return ret;
}
#endif
CELL latest;
CELL packed[4];
CELL pindex;
CELL dataList[1024];
CELL dataType[1024];
CELL dindex;
#define MAX_NAMES 1024
#define STRING_LEN 64
CELL packMode;
char najeLabels[MAX_NAMES][STRING_LEN];
CELL najePointers[MAX_NAMES];
CELL najeRefCount[MAX_NAMES];
CELL np;
CELL references[IMAGE_SIZE];
CELL pass;
char outputName[STRING_LEN];
CELL najeLookup(char *name) {
  CELL slice = -1;
  CELL n = np;
  while (n > 0) {
    n--;
    if (strcmp(najeLabels[n], name) == 0)
      slice = najePointers[n];
  }
  return slice;
}
CELL najeLookupPtr(char *name) {
  CELL slice = -1;
  CELL n = np;
  while (n > 0) {
    n--;
    if (strcmp(najeLabels[n], name) == 0)
      slice = n;
  }
  return slice;
}
void najeAddLabel(char *name, CELL slice) {
  if (pass == 1)  // labels recorded only in 1st pass
    return;
  if (najeLookup(name) == -1) {
    strcpy(najeLabels[np], name);
    najePointers[np] = slice;
    najeRefCount[np] = 0;
    np++;
  } else {
    printf("Fatal error: %s already defined\n", name);
    exit(0);
  }
}
void najeWriteMap() {
#ifdef ENABLE_MAP
  FILE *fp;
  CELL i;
  if ((fp = fopen(strcat(outputName, ".map"), "w")) == NULL) {
    printf("Unable to save the ngaImage.map!\n");
    exit(2);
  }
  for (i = 0; i < np; i++)
    fprintf(fp, "LABEL\t%s\t%d\n", najeLabels[i], najePointers[i]);
  for (i = 0; i < latest; i++) {
    if (references[i] == 0)
      fprintf(fp, "LITERAL\t%d\t%d\n", memory[i], i);
  }
  for (i = 0; i < latest; i++) {
    if (references[i] == -1)
      fprintf(fp, "POINTER\t%d\t%d\n", memory[i], i);
  }
  fclose(fp);
#else
  return;
#endif
}
void najeStore(CELL type, CELL value) {
  memory[latest] = value;
  references[latest] = type;
  latest = latest + 1;
}
void najeSync() {
  CELL i;
  if (packMode == 0)
    return;
  if (pindex == 0 && dindex == 0)
    return;
  if (pindex != 0) {
    unsigned int opcode = 0;
    opcode = packed[3];
    opcode = opcode << 8;
    opcode += packed[2];
    opcode = opcode << 8;
    opcode += packed[1];
    opcode = opcode << 8;
    opcode += packed[0];
    najeStore(2, opcode);
  }
  if (dindex != 0) {
    for (i = 0; i < dindex; i++)
      najeStore(dataType[i], dataList[i]);
  }
  pindex = 0;
  dindex = 0;
  packed[0] = 0;
  packed[1] = 0;
  packed[2] = 0;
  packed[3] = 0;
}
void najeInst(CELL opcode) {
  if (packMode == 0)
    najeStore(0, opcode);
  else {
    if (pindex == 4) {
      najeSync();
    }
    packed[pindex] = opcode;
    pindex++;
    switch (opcode) {
      case 7:
      case 8:
      case 9:
      case 10:
      case 25:
      case 26: najeSync();
               break;
      default: break;
    }
  }
}
void najeData(CELL type, CELL data) {
  if (packMode == 0)
    najeStore(type, data);
  else {
    dataList[dindex] = data;
    dataType[dindex] = type;
    dindex++;
  }
}
void najeAssemble(char *source) {
  CELL i;
  char *token;
  char *rest;
  char *ptr = source;
  char relevant[3];
  relevant[0] = 0;
  relevant[1] = 0;
  relevant[2] = 0;
  if (strlen(source) == 0)
    return;
  token = strtok_r(ptr, " ,", &rest);
  ptr = rest;
  relevant[0] = (char)token[0];
  relevant[1] = (char)token[1];
  /* Labels start with : */
  if (relevant[0] == ':') {
    najeSync();
    najeAddLabel((char *)token + 1, latest);
  }
  /* Directives start with . */
  if (relevant[0] == '.') {
    switch (relevant[1]) {
      case 'r': /* .reference */
                token = strtok_r(ptr, " ,", &rest);
                if (pass == 0) {
                  najeData(1, -9999);
                } else {
                  najeData(-1, najeLookup((char *) token));
		  najeRefCount[najeLookupPtr((char *) token)]++;
		}
                break;
      case 'c': /* .comment */
                break;
      case 'd': /* .data */
                token = strtok_r(ptr, " ,", &rest);
                najeSync();
                najeData(0, atoi(token));
                najeSync();
                break;
      case 'o': /* .output */
                token = strtok_r(ptr, " ,", &rest);
                strcpy(outputName, token);
                break;
      case 'p': /* set packed mode */
                packMode = 1;
                break;
      case 'u': /* set unpacked mode */
                najeSync();
                packMode = 0;
                break;
      case 'a': /* .allocate */
                token = strtok_r(ptr, " ,", &rest);
                i = atoi(token);
                najeSync();
                while (i > 0) {
                  najeData(0, 0);
                  i--;
                }
                najeSync();
                break;
      case 's': /* .string */
                token = strtok_r(ptr, "\n", &rest);
                i = 0;
                najeSync();
                while (i < strlen(token)) {
                  najeData(0, token[i]);
                  i++;
                }
                najeData(0, 0);
                najeSync();
                break;
    }
  }
  /* Instructions */
  if (strcmp(relevant, "no") == 0)
    najeInst(0);
  if (strcmp(relevant, "li") == 0) {
    token = strtok_r(ptr, " ,", &rest);
    najeInst(1);
    if (token[0] == '&') {
      if (pass == 0) {
        najeData(1, -9999);
      } else {
        najeData(-1, najeLookup((char *) token+1));
        najeRefCount[najeLookupPtr((char *) token+1)]++;
      }
    } else {
      najeData(0, atoi(token));
    }
  }
  if (strcmp(relevant, "du") == 0)
    najeInst(2);
  if (strcmp(relevant, "dr") == 0)
    najeInst(3);
  if (strcmp(relevant, "sw") == 0)
    najeInst(4);
  if (strcmp(relevant, "pu") == 0)
    najeInst(5);
  if (strcmp(relevant, "po") == 0)
    najeInst(6);
  if (strcmp(relevant, "ju") == 0)
    najeInst(7);
  if (strcmp(relevant, "ca") == 0)
    najeInst(8);
  if (strcmp(relevant, "cc") == 0)
    najeInst(9);
  if (strcmp(relevant, "re") == 0)
    najeInst(10);
  if (strcmp(relevant, "eq") == 0)
    najeInst(11);
  if (strcmp(relevant, "ne") == 0)
    najeInst(12);
  if (strcmp(relevant, "lt") == 0)
    najeInst(13);
  if (strcmp(relevant, "gt") == 0)
    najeInst(14);
  if (strcmp(relevant, "fe") == 0)
    najeInst(15);
  if (strcmp(relevant, "st") == 0)
    najeInst(16);
  if (strcmp(relevant, "ad") == 0)
    najeInst(17);
  if (strcmp(relevant, "su") == 0)
    najeInst(18);
  if (strcmp(relevant, "mu") == 0)
    najeInst(19);
  if (strcmp(relevant, "di") == 0)
    najeInst(20);
  if (strcmp(relevant, "an") == 0)
    najeInst(21);
  if (strcmp(relevant, "or") == 0)
    najeInst(22);
  if (strcmp(relevant, "xo") == 0)
    najeInst(23);
  if (strcmp(relevant, "sh") == 0)
    najeInst(24);
  if (strcmp(relevant, "zr") == 0)
    najeInst(25);
  if (strcmp(relevant, "en") == 0)
    najeInst(26);
}
void prepare() {
  if (pass == 0)
    np = 0;
  latest = 0;
  packMode = 1;
  strcpy(outputName, "ngaImage");
  /* assemble the standard preamble (a jump to :main) */
  najeInst(1);  /* LIT */
  najeData(0, 0);  /* placeholder */
  najeInst(7);  /* JUMP */
}
void finish() {
  CELL entry = najeLookup("main");
  memory[1] = entry;
}
void read_line(FILE *file, char *line_buffer) {
  if (file == NULL) {
    printf("Error: file pointer is null.");
    exit(1);
  }
  if (line_buffer == NULL) {
    printf("Error allocating memory for line buffer.");
    exit(1);
  }
  int ch = getc(file);
  CELL count = 0;
  while ((ch != '\n') && (ch != EOF)) {
    line_buffer[count] = ch;
    count++;
    ch = getc(file);
  }
  line_buffer[count] = '\0';
}
void process_file(char *fname) {
  char source[64000];
  FILE *fp;
  fp = fopen(fname, "r");
  if (fp == NULL)
    return;
  while (!feof(fp)) {
    read_line(fp, source);
    najeAssemble(source);
  }
  fclose(fp);
}
void save() {
  FILE *fp;
  if ((fp = fopen(outputName, "wb")) == NULL) {
    printf("Unable to save the ngaImage!\n");
    exit(2);
  }
  fwrite(&memory, sizeof(CELL), latest, fp);
  fclose(fp);
}
CELL main(int argc, char **argv) {
  CELL i;
  pass = 0;
  prepare();
    process_file(argv[1]);
    najeSync();
  finish();
  pass = 1;
  prepare();
    process_file(argv[1]);
    najeSync();
  finish();
  save();
  najeWriteMap();
#ifdef DEBUG
  printf("\nBytecode\n[");
  for (i = 0; i < latest; i++)
    printf("%d, ", memory[i]);
  printf("]\nLabels\n");
  for (i = 0; i < np; i++)
    printf("%s^%d.%d ", najeLabels[i], najePointers[i], najeRefCount[i]);
  printf("\n");
  printf("%d cells written to %s\n", latest, outputName);
#endif
  return 0;
}
